### PR TITLE
implement braintrust push for tools and prompts in python

### PR DESCRIPTION
**Pull request by sachinpad**

### CODE REVIEW

File: py/src/braintrust/cli/push.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    spec.loader.exec_module(module)
+
+    # Fetch a bundle ID to which to upload.
+    resp = proxy_conn().post_json(
+        "function/code",
+        {
+            "org_id": org_id(),
+            "runtime_context": {
+                "runtime": "python",
+                "version": "3.12",
```
Comment (ankrgyl) [inline py/src/braintrust/cli/push.py:35]: in JS i think we dynamically infer this and support a fixed set of versions (which we reject in the `/function/code` API). Should we do the same for python?

Comment (sachinpad) [reply py/src/braintrust/cli/push.py]: Done

File: py/src/braintrust/cli/push.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        "--api-key",
+        help="Specify a braintrust api key. If the parameter is not specified, the BRAINTRUST_API_KEY environment variable will be used.",
+    )
+    parser.add_argument(
+        "--org-name",
+        help="The name of a specific organization to connect to. This is useful if you belong to multiple.",
+    )
+    parser.add_argument(
+        "--app-url",
+        help="Specify a custom braintrust app url. Defaults to https://www.braintrust.dev. This is only necessary if you are using an experimental version of Braintrust",
```
Comment (ornellaaltunyan) [inline py/src/braintrust/cli/push.py:187]: add period

File: py/src/braintrust/framework2.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    slug: str
+    prompt: Any
+    tool_functions: List[Union[CodeFunction, Any]]
+    description: Optional[str]
+    id: Optional[str]
+    if_exists: Optional[str]
+
+
+class ToolBuilder:
+    """Builder to create a tool in braintrust."""
```
Comment (ornellaaltunyan) [inline py/src/braintrust/framework2.py:47]: capitalize Braintrust

File: py/src/braintrust/framework2.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            parameters=parameters,
+            returns=returns,
+            if_exists=if_exists,
+        )
+        global_.functions.append(f)
+        return f
+
+
+class PromptBuilder:
+    """Builder to create a prompt in braintrust.
```
Comment (ornellaaltunyan) [inline py/src/braintrust/framework2.py:89]: capitalize Braintrust

File: py/src/braintrust/cli/push.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+            }
+        )
+    api_conn().post_json("insert-functions", {"functions": functions})
+
+
+def build_parser(subparsers, parent_parser):
+    """Adds the parser for the push subcommand."""
+    parser = subparsers.add_parser(
+        "push",
+        help="Push code to braintrust",
```
Comment (ornellaaltunyan) [inline py/src/braintrust/cli/push.py:203]: Braintrust should be capital

File: py/src/braintrust/cli/push.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        parents=[parent_parser],
+    )
+    parser.add_argument(
+        "--api-key",
+        help="Specify a braintrust api key. If the parameter is not specified, the BRAINTRUST_API_KEY environment variable will be used.",
```
Comment (ornellaaltunyan) [inline py/src/braintrust/cli/push.py:208]: Braintrust should be capitalized

File: py/src/braintrust/cli/push.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    )
+    parser.add_argument(
+        "--org-name",
+        help="The name of a specific organization to connect to. This is useful if you belong to multiple.",
+    )
+    parser.add_argument(
+        "--app-url",
+        help="Specify a custom braintrust app url. Defaults to https://www.braintrust.dev. This is only necessary if you are using an experimental version of Braintrust.",
```
Comment (ornellaaltunyan) [inline py/src/braintrust/cli/push.py:216]: Braintrust should be capitalized

**Pull request merged**